airbyte-metrics/src/main/java/io/airbyte/metrics/DogstatsdMetricSingleton.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (c) 2021 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.metrics;
+
+import com.timgroup.statsd.NonBlockingStatsDClientBuilder;
+import com.timgroup.statsd.StatsDClient;
+import java.util.Map;
+import lombok.extern.slf4j.Slf4j;
+import org.slf4j.MDC;
+
+/**
+ * Light wrapper around the DogsStatsD client to make using the client slightly more ergonomic.
+ * <p>
+ * This class mainly exists to help Airbyte instrument/debug application on Airbyte Cloud.
+ * <p>
+ * Open source users are free to turn this on and consume the same metrics.
+ */
+@Slf4j
+public class DogstatsdMetricSingleton {
+
+  private static DogstatsdMetricSingleton instance;
+  private final StatsDClient statsDClient;
+  private final boolean instancePublish;
+
+  public DogstatsdMetricSingleton(final String appName, final boolean publish) {
+    instancePublish = publish;
+    statsDClient = new NonBlockingStatsDClientBuilder()
+        .prefix(appName)
+        .hostname(System.getenv("DD_AGENT_HOST"))
+        .port(Integer.parseInt(System.getenv("DD_DOGSTATSD_PORT")))
+        .build();
+  }
+
+  public static synchronized DogstatsdMetricSingleton getInstance() {
+    if (instance == null) {
+      throw new RuntimeException("You must initialize configuration with the initializeMonitoringServiceDaemon() method before getting an instance.");
```

Comment:
Should this say `.. with the initialize() method`?

Also, what is the downside of combining `getInstance()` and `initialize()` into the same method? The behavior would be that if `instance == null`, it initializes and returns. If `instance != null`, it just returns. That approach feels like it would be a bit more ergonomic, no?

Reply:
ah yes good catch on the error message.

this is a small thing - the main downside would be the `initialise` method has more parameters than the `getInstance` method. we'd have to combine those parameters if we combined the methods. because of this I prefer to keep these separate for now.

Reply:
@davinchia I'm curious how you are thinking about consumers of this class. If I want to use the DogstatsdMetricSingleton, how do I know if it has been initialized or not? Is it expected that a consumer would call `getInstance` wrapped with a try-catch to catch a RuntimeException in the case that it was already initialized, and call `initialize` in that case? That doesn't feel very ergonomic to me, but let me know if I'm thinking about this wrong

Reply:
Good question - I'm trying to follow the general singleton pattern where `initialize` is called in the main method, and `getInstance` is called everywhere else. Although it's possible to catch, most folks don't really catch the RTE since it's usually understood the singleton is initialised when the JVM first starts up.

I agree with you having to do `getInstance` is somewhat annoying, but this is a fairly common pattern and I don't really think it's worth trying to get rid of.

I can't see how to not have a simply method signature if we combine `initialize` and `getInstance` and I didn't want to spend more time playing with this.

Happy to discuss more/review any ideas/approaches you come up with!